### PR TITLE
CIP-66 support

### PR DESCRIPTION
**Pull request by shazarre**
Implement CIP 66 transaction serialization and parsing

### Compatibility

[CIP66 is an upcoming  ](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0066.md) Celo Transaction type for paying for gas with Tokens. 

For this implementation, a CIP-66 (AKA type 122) transaction will ONLY be created if you manually calculate and pass in `maxFeeInFeeCurrency`, `feeCurrency` and the `maxFeePerGas`  and `maxPriorityFeePerGas` fields. Otherwise if `feeCurrency` is provided  a cip64 transaction will be created else a typical 1559 transaction gets created. 

### Not implemented 

We have not implemented fetching the maxFeeInFeeToken, and right now if feeCurrency exists it will always fetch the maxFeePerGas / maxPriorityFeePerGas denominated in the FeeCurrency (cip64 style)

### CODE REVIEW

File: src/celo/parsers.ts [hunk 1]
```diff
@@ -47,6 +50,11 @@ export function parseTransaction<TSerialized extends CeloTransactionSerialized>(
       serializedTransaction as TransactionSerializedCIP64,
     ) as ParseTransactionReturnType<TSerialized>
 
+  if (serializedType === '0x7a')
+    return parseTransactionCIP66(
+      serializedTransaction as TransactionSerializedCIP66,
+    ) as ParseTransactionReturnType<TSerialized>
```
Comment (aaronmgdr) [inline src/celo/parsers.ts:59]: doesnt this need to go first? yeah. i think the test for 66 must go before the test for 64

Comment (aaronmgdr) [reply src/celo/parsers.ts:59]: oh no i was thinking of serializer. nevermind

File: src/celo/serializers.ts [hunk 2]
```diff
@@ -185,8 +235,11 @@ export function assertTransactionCIP42(
   }
 }
 
+type CIP_64_OR_CIP_66 = 'CIP-64' | 'CIP-66'
```
Comment (aaronmgdr) [inline src/celo/serializers.ts]: i think since this is a type we should use PascalCase

File: .changeset/nasty-mayflies-search.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+---
+"viem": minor
+---
+
+Add CIP-66 support for Celo
```
Comment (arthurgousset) [inline .changeset/nasty-mayflies-search.md]: Nit: Adding the URL as a hyperlink

```suggestion
Adds support for "[CIP-66: New Transaction Type: Celo Denominated Fees](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0066.md)" on Celo
```

File: .changeset/nasty-mayflies-search.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+
+https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0066.md
```
Comment (arthurgousset) [inline .changeset/nasty-mayflies-search.md]: Nit: Removes URL now a hyperlink above

```suggestion
```

File: src/celo/types.ts [hunk 5]
```diff
@@ -158,17 +181,32 @@ export type RpcTransactionRequestCIP64 = TransactionRequestBase<
...
 
+export type RpcTransactionRequestCIP66 = TransactionRequestBase<
+  Quantity,
+  Index
+> &
+  ExactPartial<FeeValuesEIP1559<Quantity>> & {
+    accessList?: AccessList | undefined
+    feeCurrency?: Address | undefined
+    maxFeeInFeeCurrency?: Quantity | undefined
+    gatewayFee?: undefined
```
Comment (arthurgousset) [inline src/celo/types.ts]: Looks like `gatewayFee` and `gatewayFeeRecipient` are still optional arguments here. 

Probably not in scope to remove them all in this PR, but might make sense to not add `gatewayFee` and `gatewayFeeRecipient` to CIP66 related types.

What do you think?

File: src/celo/formatters.ts [hunk 6]
```diff
@@ -53,8 +53,14 @@ export const formatters = {
     format(args: CeloRpcTransaction): CeloTransaction {
       const transaction = { feeCurrency: args.feeCurrency } as CeloTransaction
 
-      if (args.type === '0x7b') transaction.type = 'cip64'
-      else {
+      if (args.type === '0x7a') {
+        transaction.type = 'cip66'
+        transaction.maxFeeInFeeCurrency = args.maxFeeInFeeCurrency
```
Comment (nicolasbrugneaux) [inline src/celo/formatters.ts:67]: not sure if that's important, would `0n` if it's an explicit zero or would you rather have null?

File: src/celo/formatters.ts [hunk 7]
```diff
@@ -73,8 +79,12 @@ export const formatters = {
         feeCurrency: args.feeCurrency,
       } as CeloRpcTransactionRequest
 
-      if (isCIP64(args)) request.type = '0x7b'
-      else {
+      if (isCIP66(args)) {
+        request.type = '0x7a'
+        request.maxFeeInFeeCurrency = args.maxFeeInFeeCurrency
```
Comment (nicolasbrugneaux) [inline src/celo/formatters.ts]: no `hexToBigInt` required here?

File: src/celo/serializers.ts [hunk 8]
```diff
@@ -17,17 +17,25 @@ import type {
...
 export function serializeTransaction(
   transaction: CeloTransactionSerializable,
   signature?: Signature | undefined,
 ) {
-  if (isCIP64(transaction))
+  if (isCIP66(transaction)) {
+    return serializeTransactionCIP66(transaction, signature)
+  }
+
+  if (isCIP64(transaction)) {
```
Comment (nicolasbrugneaux) [inline src/celo/serializers.ts:38]: nit: seems like the lines below are formatted like that
```
if x
    return y
```

note the lack of `{` `}`, I don't care one way or the other but it should be consistent with viem's codebase

File: src/celo/serializers.ts [hunk 9]
```diff
@@ -126,6 +135,47 @@ function serializeTransactionCIP64(
...
+
+  const serializedTransaction = [
+    toHex(chainId),
+    nonce ? toHex(nonce) : '0x',
+    maxPriorityFeePerGas ? toHex(maxPriorityFeePerGas) : '0x',
+    maxFeePerGas ? toHex(maxFeePerGas) : '0x',
+    gas ? toHex(gas) : '0x',
+    to ?? '0x',
+    value ? toHex(value) : '0x',
+    data ?? '0x',
```
Comment (nicolasbrugneaux) [inline src/celo/serializers.ts]: nit: at this point we're defining like a billion times `0x` in different serializers. using a const could be sensible

**Review by nicolasbrugneaux — COMMENTED** [review]
looks good to me, Arthur already asked valid questions so I won't approve until his comments are addresses

File: src/celo/formatters.ts (hunk 7 above)
Comment (shazarre) [reply src/celo/formatters.ts]: good point, fixed + covered by a unit test

File: src/celo/serializers.ts (hunk 8 above)
Comment (shazarre) [reply src/celo/serializers.ts:38]: agreed, but IMHO in this case it becomes very unreadable

File: src/celo/serializers.ts (hunk 9 above)
Comment (shazarre) [reply src/celo/serializers.ts]: introduced `EMPTY_HEX_VALUE`

File: src/celo/types.ts (hunk 5 above)
Comment (shazarre) [reply src/celo/types.ts]: they're getting removed as we're removing CIP-42, need to rebase the branch

**Comment by aaronmgdr** [comment]
i think it might be smart ot not merge this into main until we actually want it to be in the viem release which would not be until closer to l2 launch

File: src/celo/formatters.ts (hunk 6 above)
Comment (aaronmgdr) [reply src/celo/formatters.ts:67]: if the type is cip66 then this will always be truthy anyway.

File: src/celo/types.ts [hunk 10]
```diff
@@ -97,18 +103,21 @@ export type CeloTransactionSerialized<
...
   RpcTransaction_<TPending> & {
     feeCurrency: Address | null
+    maxFeeInFeeCurrency: bigint | null
     gatewayFee: Hex | null
     gatewayFeeRecipient: Address | null
   }
 
 type RpcTransactionRequest = RpcTransactionRequest_ & {
   feeCurrency?: Address | undefined
+  maxFeeInFeeCurrency?: bigint | null
```
Comment (aaronmgdr) [inline src/celo/types.ts:136]: why are we saying its null? i think it should be undefined like feeCurrency is

File: src/celo/types.ts [hunk 11]
```diff
@@ -128,35 +138,62 @@ export type RpcTransactionCIP64<TPending extends boolean = boolean> = Omit<
...
     gatewayFeeRecipient?: undefined
     type: '0x7b'
   }
 
+export type RpcTransactionCIP66<TPending extends boolean = boolean> = Omit<
+  TransactionBase<Quantity, Index, TPending>,
+  'typeHex'
+> &
+  FeeValuesEIP1559<Quantity> & {
+    feeCurrency: Address | null
```
Comment (aaronmgdr) [inline src/celo/types.ts]: this should not ever be null or any falsy value

File: src/celo/types.ts [hunk 12]
```diff
@@ -128,35 +138,62 @@ export type RpcTransactionCIP64<TPending extends boolean = boolean> = Omit<
...
+    maxFeeInFeeCurrency: Quantity | null
```
Comment (aaronmgdr) [inline src/celo/types.ts]: will never be falsy.

File: src/celo/types.ts [hunk 13]
```diff
@@ -128,35 +138,62 @@ export type RpcTransactionCIP64<TPending extends boolean = boolean> = Omit<
...
   }
 
+export type RpcTransactionRequestCIP66 = TransactionRequestBase<
+  Quantity,
+  Index
+> &
+  ExactPartial<FeeValuesEIP1559<Quantity>> & {
+    accessList?: AccessList | undefined
+    feeCurrency?: Address | undefined
+    maxFeeInFeeCurrency?: Quantity | undefined
```
Comment (aaronmgdr) [inline src/celo/types.ts]: these would both always exist for cip66

File: src/celo/types.ts [hunk 14]
```diff
@@ -128,35 +138,62 @@ export type RpcTransactionCIP64<TPending extends boolean = boolean> = Omit<
...
   bigint,
   number,
   TPending
 > & {
   feeCurrency: Address | null
-  gatewayFee?: undefined
-  gatewayFeeRecipient?: undefined
+  maxFeeInFeeCurrency: bigint | null
+  gatewayFee: bigint | null
+  gatewayFeeRecipient: Address | null
```
Comment (aaronmgdr) [inline src/celo/types.ts]: this looks wrong. there is no case where both these exist on the same transaction object

**Comment by aaronmgdr** [comment]
closing as 66 has been taking off standards track

**Pull request closed without merging**